src/ray/core_worker/task_manager.h

```diff
@@ -312,25 +314,43 @@ class TaskManager : public TaskFinisherInterface, public TaskResubmissionInterfa
       for (size_t i = 0; i < num_returns; i++) {
         reconstructable_return_ids.insert(spec.ReturnId(i));
       }
-      counter.Increment({spec.GetName(), rpc::TaskStatus::PENDING_ARGS_AVAIL});
+      auto new_status =
+          std::make_tuple(spec.GetName(), rpc::TaskStatus::PENDING_ARGS_AVAIL, false);
+      counter.Increment(new_status);
+      status = new_status;
     }
 
     void SetStatus(rpc::TaskStatus new_status) {
-      counter.Swap({spec.GetName(), status}, {spec.GetName(), new_status});
-      status = new_status;
+      auto new_tuple = std::make_tuple(spec.GetName(), new_status, is_retry_);
+      counter.Swap(status, new_tuple);
+      status = new_tuple;
+    }
+
+    void MarkRetryOnFailed() {
+      // Record a separate counter increment for retries. This means that if a task
+      // is retried N times, we show it as N separate task counts.
+      counter.Increment({spec.GetName(), rpc::TaskStatus::FAILED, is_retry_});
+      is_retry_ = true;
+    }
+
+    void MarkRetryOnResubmit() {
+      // Record a separate counter increment for resubmits. This means that if a task
+      // is resubmitted N times, we show it as N separate task counts.
+      counter.Increment({spec.GetName(), rpc::TaskStatus::FINISHED, is_retry_});
```

Comment:
Why is the status FINISHED? I thought it would be pending or something like that.

Reply:
This is incrementing the metric for the "previous" attempt, which indeed should be finished. This task entry will be recording metrics for the "current" attempt from now on.

Reply:
Ah I see, maybe add a comment to explain?